ext/mbstring/libmbfl/filters/mbfilter_utf8.c

```diff
@@ -216,3 +221,106 @@ int mbfl_filt_conv_wchar_utf8(int c, mbfl_convert_filter *filter)
 
 	return 0;
 }
+
+static size_t mb_utf8_to_wchar(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state)
+{
+	unsigned char *p = *in, *e = p + *in_len;
+	uint32_t *out = buf, *limit = buf + bufsize;
+
+	while (p < e && out < limit) {
+		unsigned char c = *p++;
+
+		if (c < 0x80) {
+			*out++ = c;
+		} else if (c >= 0xC2 && c <= 0xDF) { /* 2 byte character */
+			if (p < e) {
+				char c2 = *p++;
+				if ((c2 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p--;
+				} else {
+					*out++ = ((c & 0x1F) << 6) | (c2 & 0x3F);
+				}
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+			}
+		} else if (c >= 0xE0 && c <= 0xEF) { /* 3 byte character */
+			if ((e - p) >= 2) {
+				char c2 = *p++;
+				char c3 = *p++;
+				if ((c2 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p -= 2;
+				} else if ((c3 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p--;
+				} else {
+					uint32_t decoded = ((c & 0xF) << 12) | ((c2 & 0x3F) << 6) | (c3 & 0x3F);
+					if (decoded < 0x800 || (decoded >= 0xD800 && decoded <= 0xDFFF)) {
+						*out++ = MBFL_BAD_INPUT;
+					} else {
+						*out++ = decoded;
+					}
+				}
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+				p = e; /* Remainder of string is too short to complete a 3-byte char; jump to end */
```

Comment:
Not really related to your change, but I wonder if that's the right behavior / how other libraries treat this. E.g. I could see the argument for `\xE0\x20` to result in `\U+BAD\x20`, like it would if the string did not end prematurely.

Reply:
Hmm. The legacy behavior for mbstring is to emit a single error marker if a multi-byte UTF-8 char is truncated at the end of the string.

We can change that if you want, but I think that should be a different commit.